rest_framework/views.py

```diff
@@ -62,10 +62,19 @@ def get_view_description(view, html=False):
     return description
 
 
-def set_rollback():
-    atomic_requests = connection.settings_dict.get('ATOMIC_REQUESTS', False)
-    if atomic_requests and connection.in_atomic_block:
-        transaction.set_rollback(True)
+def set_rollback(request):
+    # We need the actual view func returned by the URL resolver which gets used
+    # by Django's BaseHandler to determine `non_atomic_requests`. Be cautious
+    # when fetching it though as it won't be set when views are tested with
+    # requessts from a RequestFactory.
+    try:
+        non_atomic_requests = request.resolver_match.func._non_atomic_requests
+    except AttributeError:
+        non_atomic_requests = set()
+
+    for db in connections.all():
+        if db.settings_dict['ATOMIC_REQUESTS'] and db.alias not in non_atomic_requests:
+            transaction.set_rollback(True, using=db.alias)
```

Comment:
I'm a bit wary of this. Why is `connection.in_atomic_block` not sufficient, since it *seems* like it'd be the right thing to be doing? What does `connection.in_atomic_block` return when inside a `non_atomic_request` decorated view?

Is it possible to isolate the multi-DB fix in this PR from the `_non_atomic_requests` change in the PR, or are they tightly linked?

Reply:
> Why is `connection.in_atomic_block` not sufficient, since it seems like it'd be the right thing to be doing? What does `connection.in_atomic_block` return when inside a non_atomic_request decorated view?

`connection.in_atomic_block` means "is the default DB in a transaction?" not "is any DB in a transaction started by `ATOMIC_REQUESTS`?"

This has a number of subtleties. `connection.in_atomic_block` can return `True` in a view with `@non_atomic_requests` if there is an atomic block from another source than `ATOMIC_REQUESTS`. This is most likely to be from Django's `TestCase`, but could also be a custom view decorator or middleware for managing transactions.

More information on the specific case I had was:

* Move a client's app to `ATOMIC_REQUESTS = True` on default DB because they had a number of errors due to not using transactions
* Wrap some views with `@non_atomic_requests` because they weren't safe for ATOMIC_REQUESTS, and instead manually decorate with `@atomic` internally
* Have unit tests for those views using django's `TestCase`, which sets up two atomic blocks around tests.
* Those unit tests crash. DRF's `set_rollback()` sees `connection.in_atomic_block` is `True`, despite the transaction not coming from ATOMIC_REQUESTS. Tests checking error paths in the views cause attempt to rollback the transaction from `TestCase`, which `TestCase` also tries to rollback, which breaks the whole test run due to unbalanced transactions.

> Is it possible to isolate the multi-DB fix in this PR from the _non_atomic_requests change in the PR, or are they tightly linked?

The multi-DB fix comes "for free" because `set_rollback()` here now copies what Django does [in BaseHandler](https://github.com/django/django/blob/964dd4f4f208722d8993a35c1ff047d353cea1ea/django/core/handlers/base.py#L64). I think moving away from this is riskier than trying to make this patch focussed only on the single DB case.


I have monkey patched the implementation in this PR into my client's app to fix things there and there have been no issues for 3 months now.

Reply:
Gotcha. Is there any more graceful way we can do `connection.in_atomic_block` on a per-db basis?

Reply:
Eg, is something along these lines possible instead?...

```python
for db in connections.all():
    if db.settings_dict['ATOMIC_REQUESTS'] and db.in_atomic_block:
        transaction.set_rollback(True, using=db.alias)
```

Reply:
Yes that should work but it wouldn't fix my bug with testing the `@non_atomic_requests` views under `TestCase`

Reply:
> it wouldn't fix my bug with testing the @non_atomic_requests views under TestCase

Sorry, walk me through that. Do you mean would be broken for `@non_atomic_requests` views, or that it would be broken for *test cases of `@non_atomic_requests` views*?

Reply:
The latter (`TransactionTestCase` should still work).

Reply:
Okey dokes. So would we be able to update the PR to use the style in the comment above, and switch any test cases to `TransactionTestCase` if required?

Reply:
Done.

However this doesn't fix #6921. I will still need that patch in place for my client because their tests use `TestCase` on `@non_atomic_requests` views. I remembered while writing that commit that even `raise Http404` counts as an error on DRF ;)

Reply:
"I will still need that patch in place for my client because their tests use TestCase on @non_atomic_requests views."

Okay, that seems to fall within documented expected behavior...

https://docs.djangoproject.com/en/2.2/topics/testing/tools/#django.test.TransactionTestCase "Django’s TestCase class is a more commonly used subclass of TransactionTestCase that makes use of database transaction facilities to speed up the process of resetting the database to a known state at the beginning of each test. **A consequence of this, however, is that some database behaviors cannot be tested within a Django TestCase class**. For instance, you cannot test that a block of code is executing within a transaction, as is required when using select_for_update(). **In those cases, you should use TransactionTestCase**."